Remove pslot dir from local archive folder on CI case completion

TerrenceMcGuinness-NOAA opened the pull request:
# Description

This Pr removes **pslot** dir (with hash) from local archive folder on CI case completion and also adds `cleanup_experiment` function in the BASH utilities to clean up after a CI case runs. This change consolidates cleanup actions and adds functionality to clean the local archive folders.

Key changes:

- Modified ci/Jenkinsfile to include cleanup steps for the `DATAROOT` directory before new experiments are ran incase the pipeline is being ran for a second time.
- Updated `$HOMEgfs/ci/scripts/check_ci.sh` to use the new `cleanup_experiment` function under the BASH CI system.
- Added `cleanup_experiment `function in `$HOMEgfs/ci/scripts/utils/ci_utils.sh` to handle various cleanups including DATAROOT, ARCDIR, and ATADIR directories.
- Added get config var from EXPDIR utility  [get_config_var.py](https://github.com/NOAA-EMC/global-workflow/pull/2961/files#diff-b780733ae1d45917730364e09a6c510e79d4cc8cffad6d9020c7961c53b987bc) for getting config values in BASH

Resolves #2954, #3066 

# Type of change
- [ ] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
<!-- Choose YES or NO from each of the following and delete the other -->
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
This PR was self tested with the CI system.
![image](https://github.com/user-attachments/assets/201208bb-9610-43a7-a84a-b450f2b7262c)# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [ ] I have made corresponding changes to the system documentation if necessary

TerrenceMcGuinness-NOAA committed:
added a cleanup_experment funciton to ci_utils that cleans local archrive files

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into ci_clean_loc_arch

TerrenceMcGuinness-NOAA committed:
moved rm for EXPDIR into cleanup_expermiment function

TerrenceMcGuinness-NOAA committed:
Merge branch 'ci_clean_loc_arch' of github.com:TerrenceMcGuinness-NOAA/global-workflow into ci_clean_loc_arch

TerrenceMcGuinness-NOAA committed:
fixed bracket sytax error in ci_utils and missnamed function call in Jenkins file to cleanup experiment

TerrenceMcGuinness-NOAA committed:
fixed SC2155 in bash utils for cleanup

TerrenceMcGuinness-NOAA committed:
fixed SC2115 for var expation in path for clean experiment util

TerrenceMcGuinness-NOAA committed:
needed the OR truch for var assigment via actions in bash

WalterKolczynski-NOAA reviewed (dismissed):
Will still need to add removal of `ATARDIR`/`LOCALDIR` in a future PR.

TerrenceMcGuinness-NOAA commented:
@WalterKolczynski-NOAA Please put the specifics to those  in an Issue Task please.
Making this a draft PR until I can verify the results of the CI Pass.

TerrenceMcGuinness-NOAA committed:
fixed the evaluation of HOMDIR so that internal strification of env vars work in cleanup_experiment

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' into ci_clean_loc_arch

TerrenceMcGuinness-NOAA committed:
still needed archive tagged back onto HOMEDIR

TerrenceMcGuinness-NOAA committed:
Merge branch 'ci_clean_loc_arch' of github.com:TerrenceMcGuinness-NOAA/global-workflow into ci_clean_loc_arch

TerrenceMcGuinness-NOAA committed:
new linee at aend of ci_utils

TerrenceMcGuinness-NOAA commented:
OK, ready for merge.

WalterKolczynski-NOAA commented on ci/scripts/utils/ci_utils.sh:
```diff
@@ -155,3 +155,25 @@ function publish_logs() {
+    PSLOT_PATH="$1"
+    pslot=$(basename "${PSLOT_PATH}")
+    HOMEDIR=$(grep 'export HOMEDIR=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]"' || true) || true
+    eval HOMEDIR="${HOMEDIR}"
```
What is this `eval` doing? Isn't it already defined on the prior line? `eval` should be avoided if at all possible.

TerrenceMcGuinness-NOAA replied:
The piping left some outer quotes around some strings so that some of the variables where not resolving without it.

DavidHuber-NOAA commented on ci/scripts/utils/ci_utils.sh:
There are a few other directories that contain data, though none are very large:

```bash
${ARCHIVEDIR}/metplus_data/by_VSDB/{grid2grid,grid2obs,precip}/*/*/${pslot}
${HOME}/.rocoto/<rocoto version>/${pslot}  # These are typically empty directories on success with just a log and .lock file otherwise
```
EDIT: changed `${HOMEDIR}` to `${HOME}` for the rocoto directory.

TerrenceMcGuinness-NOAA replied:
I think is is unsafe to mess with user's Rocoto hidden state/conf directory.  I would rather not add this.

DavidHuber-NOAA replied:
Alright, that's fine.  What do you think about the `metplus_data` directories?

TerrenceMcGuinness-NOAA replied:
I do not understand how to arrive at that path for Metplus and how it is set.

DavidHuber-NOAA replied:
I'm OK with skipping those as well as they are not large and the workflow can overwrite any existing files without an issue.

WalterKolczynski-NOAA commented on ci/scripts/utils/ci_utils.sh:
```diff
@@ -155,3 +155,25 @@ function publish_logs() {
+
+    PSLOT_PATH="$1"
+    pslot=$(basename "${PSLOT_PATH}")
+    HOMEDIR=$(grep 'export HOMEDIR=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]"' || true) || true
+    eval HOMEDIR="${HOMEDIR}"
```
Try this
```suggestion
    local regex="^export HOMEDIR=(.*)"
    local lines
    mapfile -t lines < "${PSLOT_PATH}/config.base"
    local line
    for line in "${lines[@]}"; do
        if [[ "${line}" =~ ${regex} ]]; then
            HOMEDIR=$(echo "${BASH_REMATCH[1]}" | envsubst)
        fi
    done
```

TerrenceMcGuinness-NOAA replied:
Wow looks strong,  yeah I'll try that.

TerrenceMcGuinness-NOAA replied:
@DavidHuber-NOAA These archive paths designations should not be done with these one off regex injections.  Do we have a central location with an aggregation of known _archival points_?

TerrenceMcGuinness-NOAA replied:
@WalterKolczynski-NOAA I noticed that `envsubst` actually worked to resolve **$USER** without using `eval`.

TerrenceMcGuinness-NOAA replied:
@WalterKolczynski-NOAA `envsubst` was able to resolve **$USER** in **$HOMDIR** without the use of `eval`.

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into ci_clean_loc_arch

TerrenceMcGuinness-NOAA committed:
removed eval in cleanup_experiment and added envsubst at end of pipe instead

WalterKolczynski-NOAA commented on ci/scripts/utils/ci_utils.sh:
```diff
@@ -155,3 +155,24 @@ function publish_logs() {
+    PSLOT_PATH="$1"
+    pslot=$(basename "${PSLOT_PATH}")
+    HOMEDIR=$(grep 'export HOMEDIR=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]"' | envsubst || true) || true
+    ARCHIVEDIR="${HOMEDIR}/archive"
```
I think this should be reading in `$ARCDIR` from `config.base`, similar to how you are reading in `$HOMEDIR`. Note that the default ARCDIR is `${NOSCRUB}/archive`, not `${HOMEDIR}/archive`. `$ARCDIR` already includes the pslot, so it won't need to be added to the next line.

@WalterKolczynski-NOAA On the other hand `export NOSCRUB="$HOMEDIR"` is also in `confg.base`

TerrenceMcGuinness-NOAA replied:
Yeah, we should have lead with that.

TerrenceMcGuinness-NOAA replied:
Ok updated using ARCDIR.

TerrenceMcGuinness-NOAA replied:
Test failed because **ARCDIR** has an unresolved sub-string `${NOSCRUB}` derived from `${HOMEDIR}`

TerrenceMcGuinness-NOAA replied:
@WalterKolczynski-NOAA David Huber came up with idea of scraping the archive log files in **COMROOT**:

```
mterry (hercules-login-4) COMROOT $ echo $COMROOT
C96_S2SWA_gefs_replay_ics_32b8cd0c
mterry (hercules-login-4) COMROOT $ sed -n "s/file_utils  : Copied [[:graph:]]* to \([[:graph:]]*\)/\1/p" "${COMROOT}"/logs/*/*arch.log | awk '{print $NF}' | uniq
/work2/noaa/stmp/CI/HERCULES/2961/RUNTESTS/COMROOT/C96_S2SWA_gefs_replay_ics_32b8cd0c/gefsarch

mterry (hercules-login-4) COMROOT $ ls /work2/noaa/stmp/CI/HERCULES/2961/RUNTESTS/COMROOT/C96_S2SWA_gefs_replay_ics_32b8cd0c/gefsarch
gefs.t00z.mean.pres_.1p00.f006.grib2  gefs.t00z.mean.pres_.1p00.f012.grib2  gefs.t00z.mean.pres_.1p00.f018.grib2  gefs.t00z.mean.pres_.1p00.f024.grib2	gefs.t00z.mean.pres_.1p00.f030.grib2  gefs.t00z.mean.pres_.1p00.f036.grib2  gefs.t00z.mean.pres_.1p00.f042.grib2  gefs.t00z.mean.pres_.1p00.f048.grib2
```
I'll get that in tomorrow as another proposed solution.

WalterKolczynski-NOAA replied:
Instead of that, why not just feed the value of `$NOSCRUB` in?
```
ARCHIVEDIR=$(grep 'export ARCDIR=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]' | NOSCRUB="${HOMEDIR}" envsubst || true) || true
```

Or if we want to be more rigorous
```
pslot=$(basename "${PSLOT_PATH}")
HOMEDIR=$(grep 'export HOMEDIR=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]' | envsubst || true) || true
NOSCRUB=$(grep 'export NOSCRUB=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]' | HOMEDIR="${HOMEDIR}" envsubst || true) || true
ARCHIVEDIR=$(grep 'export ARCDIR=' "${PSLOT_PATH}/config.base" | cut -d'=' -f2 | tr -d '[:space:]' | NOSCRUB="${NOSCRUB}" envsubst || true) || true
```

Or we could just cut to the chase and source `config.base` and get everything.

TerrenceMcGuinness-NOAA replied:
IMHO "reconstructing" paths from hidden knowledge of their specificity such as interim archive placements is a non-systematic and haphazard approach resulting in a one-off paradigm.  If we really want to be rigorous I would propose a transformation from logs to a systematic and intent-full pedigree framework that could be leveraged for a requirement such as this.

TerrenceMcGuinness-NOAA committed:
updated archival directory to ARCDIR from config.base in cleanup_experiment

TerrenceMcGuinness-NOAA committed:
forgot to clean up EXPDIR replacement in cleanup_experment

TerrenceMcGuinness-NOAA committed:
forgot to clean up EXPDIR replacement in cleanup_experment still

DavidHuber-NOAA commented on ci/scripts/utils/ci_utils.sh:
```diff
@@ -155,3 +155,22 @@ function publish_logs() {
+    ARCDIR=$(grep 'export ARCDIR=' "${EXPDIR}/config.base" | cut -d'=' -f2 | tr -d '[:space:]"' | envsubst || true) || true
+
+    rm -Rf "${ARCDIR:?}"
+    rm -Rf "${EXPDIR}"
+    rm -Rf "${EXPDIR}/../COMROOT/${pslot}"
```
I don't think you can do a relative path via a non-existent path, so the `COMROOT/${pslot}` directory probably needs to be deleted first:

```suggestion
    rm -Rf "${EXPDIR}/../COMROOT/${pslot}"
    rm -Rf "${EXPDIR}"
```

TerrenceMcGuinness-NOAA replied:
of course, duh ◔_◔

TerrenceMcGuinness-NOAA replied:
of course, duh ◔_◔

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/ci_utils.sh

of course, duh ◔_◔

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

DavidHuber-NOAA reviewed (dismissed):
Looks good to me!

DavidHuber-NOAA commented:
@TerrenceMcGuinness-NOAA Let me know when you have been able to test this and I will go ahead and merge it.

TerrenceMcGuinness-NOAA committed:
added get_config_var util to simplify cleanup_experment

TerrenceMcGuinness-NOAA committed:
removed debug outputs in cleanup_experiment

TerrenceMcGuinness-NOAA committed:
added next step to use get_config_var to remove previous DATAROOT data when startig new experimentsa

TerrenceMcGuinness-NOAA committed:
fix conficts and merged develop back in

TerrenceMcGuinness-NOAA committed:
updated Jenkinsfile to use clean experiment and to also clean DATADIR in advance of running a experiment for the case that it had already ran

TerrenceMcGuinness-NOAA committed:
updating correct ATARDIR variable in cleaup

TerrenceMcGuinness-NOAA committed:
added true for shell norms

DavidHuber-NOAA commented on ci/scripts/utils/ci_utils.sh:
```diff
@@ -155,3 +154,24 @@ function publish_logs() {
+    pslot=$(basename "${EXPDIR}")
+
+    # Use the Python utility to get the required variables
+    read -r ARCDIR ATARDIR DATAROOT < <("${HOMEgfs}/ci/scripts/utils/get_config_var.py" ARCDIR ATARDIR DATAROOT "${EXPDIR}") || true
+
+    rm -Rf "${DATAROOT:?}"
+    rm -Rf "${ARCDIR:?}"
+    rm -Rf "${ATARDIR:?}"
+    rm -Rf "${EXPDIR}/../COMROOT/${pslot}"
```
You could also read `COMROOT` from `config.base`:

```suggestion
    read -r ARCDIR ATARDIR DATAROOT COMROOT < <("${HOMEgfs}/ci/scripts/utils/get_config_var.py" ARCDIR ATARDIR DATAROOT COMROOT "${EXPDIR}") || true

    rm -Rf "${DATAROOT:?}"
    rm -Rf "${ARCDIR:?}"
    rm -Rf "${ATARDIR:?}"
    rm -Rf "${COMROOT:?}"
```

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/ci_utils.sh

Of course, love it.

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

DavidHuber-NOAA commented on ci/scripts/utils/ci_utils.sh:
```diff
@@ -155,3 +154,24 @@ function publish_logs() {
+    local pslot
+    local ARCDIR
+    local ATARDIR
+    local DATAROOT
```
Sorry, missed one spot:

```suggestion
    local DATAROOT
    local COMROOT
```

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/ci_utils.sh

oh sure, yes. Notice also we do the DATAROOT remove independent of the cleanup_experiment before it is ran again in Jenkins.

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

DavidHuber-NOAA commented on ci/scripts/utils/get_config_var.py:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+
+import os
+import sys
```
```suggestion
```

DavidHuber-NOAA commented on ci/scripts/utils/get_config_var.py:
```diff
@@ -0,0 +1,44 @@
+import sys
+import argparse
+from wxflow import Configuration
+
```
```suggestion


```

DavidHuber-NOAA commented on ci/scripts/utils/get_config_var.py:
```diff
@@ -0,0 +1,44 @@
+    config = Configuration(config_dir)
+    config_data = config.parse_config(config_file)
+    return [config_data[var_name] for var_name in var_names]
+
```
```suggestion


```

DavidHuber-NOAA reviewed (changes requested):
Fixes for the pynorms complaints.

DavidHuber-NOAA commented on ci/scripts/utils/get_config_var.py:
```diff
@@ -0,0 +1,44 @@
+def get_config_vars(var_names, config_path):
+    """
+    GET_CONFIG_VARS Get configuration variables from a config file or directory.
+    
```
```suggestion

```

DavidHuber-NOAA commented on ci/scripts/utils/get_config_var.py:
```diff
@@ -0,0 +1,44 @@
+    Parameters:
+    var_names (list of str): The names of the configuration variables to retrieve.
+    config_path (str): The path to the configuration file or directory.
+    
```
```suggestion

```

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/get_config_var.py

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/get_config_var.py

que?

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/get_config_var.py

que que

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/get_config_var.py

que que que

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/get_config_var.py

que que que que

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
fixed name of get_config_var in pipeline (left off the py extention)

TerrenceMcGuinness-NOAA committed:
DATAROOT is not a config.base value but is construced from STMP in tasks.py

TerrenceMcGuinness-NOAA committed:
updated Jenkins file with correct RUNDIRS string (had RUNDIR as a bug)

TerrenceMcGuinness-NOAA commented:
Changes moved to [PR 3076](https://github.com/NOAA-EMC/global-workflow/pull/3076)
